From 5d974657c879ad535cf1f097248e94329ec53f07 Mon Sep 17 00:00:00 2001
From: tronzhang <zhangzhaochuang@huawei.com>
Date: Thu, 23 Jan 2025 10:18:53 +0800
Subject: [PATCH] local

---
 setup.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/setup.py b/setup.py
index 61d2d71..d1ffd08 100644
--- a/setup.py
+++ b/setup.py
@@ -537,7 +537,7 @@ def get_requirements() -> List[str]:
         return resolved_requirements
 
     if _no_device():
-        requirements = _read_requirements("requirements-cuda.txt")
+        requirements = _read_requirements("requirements-common.txt")
     elif _is_cuda():
         requirements = _read_requirements("requirements-cuda.txt")
         cuda_major, cuda_minor = torch.version.cuda.split(".")
-- 
2.43.0

